File: 2058-find-minimum-and-maximum-number-of-nodes-between-critical-points/Solution.py
class Solution:
    def nodesBetweenCriticalPoints(self, head):
        first = -1
        previous = -1
        min_distance = float("inf")

        prev = head
        curr = head.next

        position = 1

        while curr.next:
            next_node = curr.next

            if ((curr.val > prev.val and curr.val > next_node.val) or
                (curr.val < prev.val and curr.val < next_node.val)):

                if first == -1:
                    first = position
                else:
                    min_distance = min(
                        min_distance,
                        position - previous
                    )

                previous = position

            prev = curr
            curr = next_node
            position += 1

        if first == -1 or previous == first:
            return [-1, -1]

        max_distance = previous - first

        return [min_distance, max_distance]

File: 2058-find-minimum-and-maximum-number-of-nodes-between-critical-points/Solution.java
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        int first = -1;
        int previous = -1;
        int minDistance = Integer.MAX_VALUE;

        ListNode prev = head;
        ListNode curr = head.next;

        int position = 1;

        while (curr.next != null) {
            ListNode next = curr.next;

            if ((curr.val > prev.val && curr.val > next.val) ||
                (curr.val < prev.val && curr.val < next.val)) {

                if (first == -1) {
                    first = position;
                } else {
                    minDistance = Math.min(minDistance, position - previous);
                }

                previous = position;
            }

            prev = curr;
            curr = next;
            position++;
        }

        if (first == -1 || previous == first) {
            return new int[]{-1, -1};
        }

        int maxDistance = previous - first;

        return new int[]{minDistance, maxDistance};
    }
}
